File: go/signature_test.go
package plutus

import (
	"errors"
	"net/http"
	"strings"
	"testing"
)

func TestRequestSignatureVectors(t *testing.T) {
	for _, vec := range vectors(t).Vectors.RequestSignature {
		vec := vec
		t.Run(vec.ID, func(t *testing.T) {
			req := CanonicalRequest{
				Method:         vec.Request.Method,
				ExternalPath:   vec.Request.ExternalPath,
				RawQuery:       str(vec.Request.QueryString),
				Timestamp:      vec.Request.Timestamp,
				Nonce:          vec.Request.Nonce,
				APIVersion:     vec.Request.APIVersion,
				IdempotencyKey: str(vec.Request.IdempotencyKey),
				Body:           bodyBytes(vec.Request.Body),
			}

			canonicalQuery, err := CanonicalizeQuery(req.RawQuery)
			if err != nil {
				t.Fatalf("canonicalize query: %v", err)
			}
			if canonicalQuery != vec.CanonicalQuery {
				t.Errorf("canonical query = %q, want %q", canonicalQuery, vec.CanonicalQuery)
			}
			if got := ForcesEmptyBodyDigest(req.Method); got != vec.ForcedEmptyBody {
				t.Errorf("forced empty body = %v, want %v", got, vec.ForcedEmptyBody)
			}
			if got := BodyDigestHex(req.Method, req.Body); got != vec.BodyHash {
				t.Errorf("body digest = %s, want %s", got, vec.BodyHash)
			}

			canonical, err := req.CanonicalString()
			if err != nil {
				t.Fatalf("canonical string: %v", err)
			}
			lines := strings.Split(canonical, "\n")
			if len(lines) != len(vec.CanonicalStringLines) {
				t.Fatalf("canonical string has %d lines, want %d", len(lines), len(vec.CanonicalStringLines))
			}
			for i, want := range vec.CanonicalStringLines {
				if lines[i] != want {
					t.Errorf("canonical line %d = %q, want %q", i+1, lines[i], want)
				}
			}
			if canonical != vec.CanonicalString {
				t.Fatalf("canonical string mismatch:\n got %q\nwant %q", canonical, vec.CanonicalString)
			}
			if got := SHA256Hex([]byte(canonical)); got != vec.RequestCanonicalSHA256 {
				t.Errorf("request canonical digest = %s, want %s", got, vec.RequestCanonicalSHA256)
			}

			// 用 merchant_auth 公钥验证向量签名。
			if err := VerifyCanonicalString(publicKey(t, vec.SigningKey), canonical, vec.Signature); err != nil {
				t.Fatalf("verify vector signature: %v", err)
			}

			// PKCS#1 v1.5 是确定性的: 重新签名必须与向量逐字节相等。
			signer := NewSignerWithKey(vec.Headers[HeaderAPIKey], privateKey(t, vec.SigningKey))
			signed, err := signer.Sign(req)
			if err != nil {
				t.Fatalf("sign: %v", err)
			}
			if signed.Signature != vec.Signature {
				t.Fatalf("signature mismatch:\n got %s\nwant %s", signed.Signature, vec.Signature)
			}
			if signed.CanonicalDigestHex != vec.RequestCanonicalSHA256 {
				t.Errorf("canonical digest = %s, want %s", signed.CanonicalDigestHex, vec.RequestCanonicalSHA256)
			}
			for name, want := range vec.Headers {
				if got := signed.Headers[name]; got != want {
					t.Errorf("header %s = %q, want %q", name, got, want)
				}
			}
			if len(signed.Headers) != len(vec.Headers) {
				t.Errorf("produced %d headers, vector has %d: %v", len(signed.Headers), len(vec.Headers), signed.Headers)
			}
		})
	}
}

func TestRequestSignatureRejectsTamperedCanonical(t *testing.T) {
	vec := vectors(t).Vectors.RequestSignature[3]
	pub := publicKey(t, vec.SigningKey)
	tampered := strings.Replace(vec.CanonicalString, "POST", "PUT", 1)
	if err := VerifyCanonicalString(pub, tampered, vec.Signature); err == nil {
		t.Fatal("tampered canonical string must fail verification")
	}
	// 幂等键行被清空后同样必须失败。
	withoutIdem := strings.Replace(vec.CanonicalString, "idem-vector-0000000004", "", 1)
	if err := VerifyCanonicalString(pub, withoutIdem, vec.Signature); err == nil {
		t.Fatal("canonical string without idempotency key must fail verification")
	}
}

func TestSignerValidatesNonceAndTimestamp(t *testing.T) {
	signer := NewSignerWithKey("apk_test", privateKey(t, "merchant_auth"))
	base := CanonicalRequest{
		Method:       "GET",
		ExternalPath: "/card-products/cards/page",
		Timestamp:    "1755600000000",
		Nonce:        "nonce-vector-0000000001",
		APIVersion:   DefaultAPIVersion,
	}

	bad := base
	bad.Nonce = "c2hvcnQ="
	if _, err := signer.Sign(bad); !errors.Is(err, ErrInvalidNonce) {
		t.Fatalf("base64 nonce must be rejected, got %v", err)
	}

	bad = base
	bad.Nonce = "short"
	if _, err := signer.Sign(bad); !errors.Is(err, ErrInvalidNonce) {
		t.Fatalf("short nonce must be rejected, got %v", err)
	}

	bad = base
	bad.Timestamp = "1755600000"
	if _, err := signer.Sign(bad); !errors.Is(err, ErrInvalidTimestamp) {
		t.Fatalf("second-precision timestamp must be rejected, got %v", err)
	}

	bad = base
	bad.RawQuery = "q=a+b"
	if _, err := signer.Sign(bad); !errors.Is(err, ErrInvalidQuery) {
		t.Fatalf("raw '+' in query must be rejected, got %v", err)
	}
}

func TestNewNonceIsValid(t *testing.T) {
	seen := make(map[string]bool)
	for i := 0; i < 100; i++ {
		nonce, err := NewNonce()
		if err != nil {
			t.Fatalf("NewNonce: %v", err)
		}
		if err := ValidateNonce(nonce); err != nil {
			t.Fatalf("generated nonce %q is invalid: %v", nonce, err)
		}
		if seen[nonce] {
			t.Fatalf("duplicate nonce %q", nonce)
		}
		seen[nonce] = true
	}
	if err := ValidateNonce("f81d4fae-7dec-11d0-a765-00a0c91e6bf6"); err != nil {
		t.Fatalf("hyphenated UUID must be a valid nonce: %v", err)
	}
}

func responseHeader(vec responseSignatureVector) http.Header {
	header := http.Header{}
	for name, value := range vec.Headers {
		header.Set(name, value)
	}
	header.Set("Content-Type", vec.ContentType)
	return header
}

func TestResponseSignatureVectors(t *testing.T) {
	requestByID := map[string]requestSignatureVector{}
	for _, req := range vectors(t).Vectors.RequestSignature {
		requestByID[req.ID] = req
	}

	for _, vec := range vectors(t).Vectors.ResponseSignature {
		vec := vec
		t.Run(vec.ID, func(t *testing.T) {
			// 绑定摘要必须来自本地重算的请求规范串, 不从响应头读取。
			reqVec, ok := requestByID[vec.RequestVectorID]
			if !ok {
				t.Fatalf("unknown request vector %q", vec.RequestVectorID)
			}
			if got := SHA256Hex([]byte(reqVec.CanonicalString)); got != vec.RequestCanonicalSHA256 {
				t.Fatalf("locally computed binding digest = %s, want %s", got, vec.RequestCanonicalSHA256)
			}

			body := []byte(vec.ResponseBody)
			if got := SHA256Hex(body); got != vec.ResponseBodyHash {
				t.Errorf("response body digest = %s, want %s", got, vec.ResponseBodyHash)
			}

			canonical := CanonicalResponse{
				RequestCanonicalSHA256: vec.RequestCanonicalSHA256,
				APIVersion:             vec.APIVersion,
				ExternalPath:           vec.ExternalPath,
				OperationID:            str(vec.OperationID),
				RequestID:              str(vec.RequestID),
				HTTPStatus:             vec.HTTPStatus,
				ContentType:            vec.ContentType,
				ResponseTimestamp:      vec.ResponseTimestamp,
				Body:                   body,
			}.CanonicalString()
			lines := strings.Split(canonical, "\n")
			if len(lines) != len(vec.CanonicalStringLines) {
				t.Fatalf("canonical string has %d lines, want %d", len(lines), len(vec.CanonicalStringLines))
			}
			for i, want := range vec.CanonicalStringLines {
				if lines[i] != want {
					t.Errorf("canonical line %d = %q, want %q", i+1, lines[i], want)
				}
			}
			if canonical != vec.CanonicalString {
				t.Fatalf("canonical string mismatch:\n got %q\nwant %q", canonical, vec.CanonicalString)
			}
			if err := VerifyCanonicalString(publicKey(t, vec.SigningKey), canonical, vec.Signature); err != nil {
				t.Fatalf("verify vector signature: %v", err)
			}

			verifier := NewResponseVerifierWithKey(publicKey(t, vec.SigningKey))
			binding := ResponseBinding{
				RequestCanonicalSHA256: vec.RequestCanonicalSHA256,
				APIVersion:             vec.APIVersion,
				ExternalPath:           vec.ExternalPath,
			}
			header := responseHeader(vec)
			if err := verifier.Verify(binding, vec.HTTPStatus, header, body); err != nil {
				t.Fatalf("ResponseVerifier.Verify: %v", err)
			}

			// 负向: 篡改响应体、Content-Type 归一化、绑定摘要错误都必须失败。
			if err := verifier.Verify(binding, vec.HTTPStatus, header, append(body, ' ')); !errors.Is(err, ErrResponseSignatureInvalid) {
				t.Errorf("tampered body: got %v, want ErrResponseSignatureInvalid", err)
			}
			normalized := responseHeader(vec)
			normalized.Set("Content-Type", strings.ToLower(strings.ReplaceAll(vec.ContentType, ";charset=UTF-8", "")))
			if vec.ContentType != normalized.Get("Content-Type") {
				if err := verifier.Verify(binding, vec.HTTPStatus, normalized, body); !errors.Is(err, ErrResponseSignatureInvalid) {
					t.Errorf("normalized content-type: got %v, want ErrResponseSignatureInvalid", err)
				}
			}
			wrongBinding := binding
			wrongBinding.RequestCanonicalSHA256 = strings.Repeat("0", 64)
			if err := verifier.Verify(wrongBinding, vec.HTTPStatus, header, body); !errors.Is(err, ErrResponseSignatureInvalid) {
				t.Errorf("wrong binding digest: got %v, want ErrResponseSignatureInvalid", err)
			}
			missing := responseHeader(vec)
			missing.Del(HeaderResponseSignature)
			if err := verifier.Verify(binding, vec.HTTPStatus, missing, body); !errors.Is(err, ErrResponseSignatureMissing) {
				t.Errorf("missing signature: got %v, want ErrResponseSignatureMissing", err)
			}
		})
	}
}

// TestParseAPIResponseFromVectors 只覆盖响应头到 APIResponse 的解析与原始字节保留。
//
// 向量文件的 body 是为签名字节固定下来的历史形态, 不代表当前的统一响应包络,
// 因此这里不做成功/失败判定断言 —— 成功判定用本地 fixture 覆盖,
// 见 envelope_success_test.go 的 TestEnvelopeSuccessDecision。
func TestParseAPIResponseFromVectors(t *testing.T) {
	for _, vec := range vectors(t).Vectors.ResponseSignature {
		resp := ParseAPIResponse(vec.HTTPStatus, responseHeader(vec), []byte(vec.ResponseBody))
		if resp.RequestID != str(vec.RequestID) {
			t.Errorf("%s: request id = %q, want %q", vec.ID, resp.RequestID, str(vec.RequestID))
		}
		if resp.OperationID != str(vec.OperationID) {
			t.Errorf("%s: operation id = %q, want %q", vec.ID, resp.OperationID, str(vec.OperationID))
		}
		if resp.HTTPStatus != vec.HTTPStatus {
			t.Errorf("%s: http status = %d, want %d", vec.ID, resp.HTTPStatus, vec.HTTPStatus)
		}
		if string(resp.Body) != vec.ResponseBody {
			t.Errorf("%s: raw body was not preserved verbatim", vec.ID)
		}
	}
}

File: node/test/signature.test.ts
import { createHash } from 'node:crypto';
import { describe, expect, it } from 'vitest';
import {
  RequestSigner,
  ResponseVerifier,
  PlutusSignatureError,
  buildRequestCanonicalString,
  buildResponseCanonicalString,
  bodyDigestHex,
  canonicalizeQuery,
  requestCanonicalSha256,
  signedBodyDigest,
  verifyCanonicalSignature,
} from '../src/index.js';
import { key, vectors } from './vectors.js';

const merchantAuth = key('merchant_auth');
const platformAuth = key('platform_auth');

describe('requestSignature 向量组', () => {
  it('向量数量为 7', () => {
    expect(vectors.vectors.requestSignature).toHaveLength(7);
  });

  for (const vector of vectors.vectors.requestSignature) {
    describe(`${vector.id}: ${vector.description}`, () => {
      const req = vector.request;
      const bodyBytes = req.body === null ? null : Buffer.from(req.body, 'utf8');

      it('query 规范化结果一致', () => {
        expect(canonicalizeQuery(req.queryString)).toBe(vector.canonicalQuery);
      });

      it('body 摘要与强制空体规则一致', () => {
        expect(signedBodyDigest(req.method, bodyBytes)).toBe(vector.bodyHash);
        expect(vector.forcedEmptyBody).toBe(['GET', 'HEAD', 'DELETE'].includes(req.method));
      });

      it('规范串逐行与整串比对', () => {
        const canonical = buildRequestCanonicalString({
          method: req.method,
          externalPath: req.externalPath,
          canonicalQuery: canonicalizeQuery(req.queryString),
          timestamp: req.timestamp,
          nonce: req.nonce,
          apiVersion: req.apiVersion,
          idempotencyKey: req.idempotencyKey,
          bodyDigest: signedBodyDigest(req.method, bodyBytes),
        });
        expect(canonical.split('\n')).toEqual(vector.canonicalStringLines);
        expect(canonical).toBe(vector.canonicalString);
      });

      it('请求绑定摘要一致', () => {
        expect(requestCanonicalSha256(vector.canonicalString)).toBe(vector.requestCanonicalSha256);
      });

      it('向量签名可用 merchant_auth 公钥验证', () => {
        expect(verifyCanonicalSignature(vector.canonicalString, vector.signature, merchantAuth.publicKeyPem)).toBe(
          true,
        );
      });

      it('RequestSigner 重签结果与向量逐字节相等(PKCS#1 v1.5 确定性)', () => {
        const signer = new RequestSigner({
          apiKey: vector.headers['X-Api-Key'] as string,
          merchantAuthPrivateKey: merchantAuth.privateKeyPem,
          apiVersion: req.apiVersion,
        });
        const signed = signer.sign({
          method: req.method,
          path: req.externalPath,
          query: req.queryString,
          body: bodyBytes,
          timestamp: req.timestamp,
          nonce: req.nonce,
          idempotencyKey: req.idempotencyKey,
        });
        expect(signed.canonicalString).toBe(vector.canonicalString);
        expect(signed.signature).toBe(vector.signature);
        expect(signed.requestCanonicalSha256).toBe(vector.requestCanonicalSha256);
        expect(signed.headers).toEqual(vector.headers);
      });
    });
  }
});

describe('requestSignature 负向', () => {
  const vector = vectors.vectors.requestSignature.find((v) => v.id === 'rq-04-post-body-idempotency');

  it('篡改规范串任意一行都会验签失败', () => {
    const lines = [...(vector as NonNullable<typeof vector>).canonicalStringLines];
    for (let i = 0; i < lines.length; i += 1) {
      const mutated = [...lines];
      mutated[i] = `${mutated[i]}x`;
      expect(
        verifyCanonicalSignature(
          mutated.join('\n'),
          (vector as NonNullable<typeof vector>).signature,
          merchantAuth.publicKeyPem,
        ),
      ).toBe(false);
    }
  });

  it('用另一把公钥验签必然失败', () => {
    expect(
      verifyCanonicalSignature(
        (vector as NonNullable<typeof vector>).canonicalString,
        (vector as NonNullable<typeof vector>).signature,
        platformAuth.publicKeyPem,
      ),
    ).toBe(false);
  });

  it('省略幂等键会改变规范串第 7 行,导致签名不同', () => {
    const signer = new RequestSigner({
      apiVersion: '1',
      apiKey: 'apk_vector_0001',
      merchantAuthPrivateKey: merchantAuth.privateKeyPem,
    });
    const v = vector as NonNullable<typeof vector>;
    const signed = signer.sign({
      method: v.request.method,
      path: v.request.externalPath,
      body: Buffer.from(v.request.body as string, 'utf8'),
      timestamp: v.request.timestamp,
      nonce: v.request.nonce,
    });
    expect(signed.signature).not.toBe(v.signature);
    expect(signed.headers['X-Idempotency-Key']).toBeUndefined();
  });

  it('nonce 不合规时本地抛错', () => {
    const signer = new RequestSigner({
      apiVersion: '1',
      apiKey: 'apk_vector_0001',
      merchantAuthPrivateKey: merchantAuth.privateKeyPem,
    });
    expect(() => signer.sign({ method: 'GET', path: '/x', nonce: 'short+/=' })).toThrowError(
      /nonce must match/,
    );
  });

  it('秒级时间戳不会被自动纠正,但格式非数字时抛错', () => {
    const signer = new RequestSigner({
      apiVersion: '1',
      apiKey: 'apk_vector_0001',
      merchantAuthPrivateKey: merchantAuth.privateKeyPem,
    });
    expect(() => signer.sign({ method: 'GET', path: '/x', timestamp: '2026-08-30T00:00:00Z' })).toThrowError(
      /Unix milliseconds/,
    );
  });

  it('默认生成的 nonce 与时间戳满足协议约束', () => {
    const signer = new RequestSigner({
      apiVersion: '1',
      apiKey: 'apk_vector_0001',
      merchantAuthPrivateKey: merchantAuth.privateKeyPem,
    });
    const signed = signer.sign({ method: 'GET', path: '/card-products/cards/page' });
    expect(signed.nonce).toMatch(/^[A-Za-z0-9._~-]{16,128}$/);
    expect(signed.timestamp).toMatch(/^\d{13}$/);
    expect(signed.headers['X-Signature-Algorithm']).toBe('RSA-SHA256');
  });
});

describe('responseSignature 向量组', () => {
  it('向量数量为 3', () => {
    expect(vectors.vectors.responseSignature).toHaveLength(3);
  });

  for (const vector of vectors.vectors.responseSignature) {
    describe(`${vector.id}: ${vector.description}`, () => {
      const bodyBytes = Buffer.from(vector.responseBody, 'utf8');

      it('绑定摘要来自关联的请求向量,而非响应头', () => {
        const requestVector = vectors.vectors.requestSignature.find((v) => v.id === vector.requestVectorId);
        expect(requestVector).toBeDefined();
        expect(requestCanonicalSha256((requestVector as NonNullable<typeof requestVector>).canonicalString)).toBe(
          vector.requestCanonicalSha256,
        );
      });

      it('响应体摘要一致', () => {
        expect(bodyDigestHex(bodyBytes)).toBe(vector.responseBodyHash);
      });

      it('10 行规范串逐行与整串比对', () => {
        const canonical = buildResponseCanonicalString({
          requestCanonicalSha256: vector.requestCanonicalSha256,
          apiVersion: vector.apiVersion,
          externalPath: vector.externalPath,
          operationId: vector.operationId,
          requestId: vector.requestId,
          httpStatus: vector.httpStatus,
          contentType: vector.contentType,
          responseTimestamp: vector.responseTimestamp,
          responseBodyDigest: vector.responseBodyHash,
        });
        expect(canonical.split('\n')).toEqual(vector.canonicalStringLines);
        expect(canonical).toBe(vector.canonicalString);
      });

      it('ResponseVerifier 通过 platform_auth 公钥验签', () => {
        const verifier = new ResponseVerifier({ platformAuthPublicKey: platformAuth.publicKeyPem });
        const headers: Record<string, string> = {
          'X-Response-Signature': vector.signature,
          'X-Response-Timestamp': vector.responseTimestamp,
          'X-Response-Signature-Algorithm': 'RSA-SHA256',
        };
        if (vector.contentType) headers['Content-Type'] = vector.contentType;
        if (vector.requestId) headers['X-Request-Id'] = vector.requestId;
        if (vector.operationId) headers['X-Operation-Id'] = vector.operationId;

        const result = verifier.verify({
          requestCanonicalSha256: vector.requestCanonicalSha256,
          apiVersion: vector.apiVersion,
          externalPath: vector.externalPath,
          status: vector.httpStatus,
          headers,
          body: bodyBytes,
        });
        expect(result.verified).toBe(true);
        expect(result.canonicalString).toBe(vector.canonicalString);
      });
    });
  }
});

describe('responseSignature 负向', () => {
  const vector = vectors.vectors.responseSignature[0] as NonNullable<
    (typeof vectors.vectors.responseSignature)[0]
  >;
  const bodyBytes = Buffer.from(vector.responseBody, 'utf8');

  function verifyWith(overrides: Record<string, unknown>): void {
    const verifier = new ResponseVerifier({ platformAuthPublicKey: platformAuth.publicKeyPem });
    verifier.verify({
      requestCanonicalSha256: vector.requestCanonicalSha256,
      apiVersion: vector.apiVersion,
      externalPath: vector.externalPath,
      status: vector.httpStatus,
      headers: {
        'X-Response-Signature': vector.signature,
        'X-Response-Timestamp': vector.responseTimestamp,
        'Content-Type': vector.contentType as string,
        'X-Request-Id': vector.requestId as string,
      },
      body: bodyBytes,
      ...overrides,
    });
  }

  it('绑定摘要不匹配(响应被挪用到别的请求)时抛错', () => {
    expect(() => verifyWith({ requestCanonicalSha256: 'f'.repeat(64) })).toThrowError(PlutusSignatureError);
  });

  it('Content-Type 被归一化后验签失败', () => {
    expect(() =>
      verifyWith({
        headers: {
          'X-Response-Signature': vector.signature,
          'X-Response-Timestamp': vector.responseTimestamp,
          'Content-Type': 'application/json;charset=UTF-8',
          'X-Request-Id': vector.requestId as string,
        },
      }),
    ).toThrowError(PlutusSignatureError);
  });

  it('响应体被改动一个字节即验签失败', () => {
    const tampered = Buffer.from(`${vector.responseBody} `, 'utf8');
    expect(() => verifyWith({ body: tampered })).toThrowError(PlutusSignatureError);
  });

  it('状态码不同即验签失败', () => {
    expect(() => verifyWith({ status: 200 === vector.httpStatus ? 201 : 200 })).toThrowError(
      PlutusSignatureError,
    );
  });

  it('默认策略:2xx 缺签名头抛错,非 2xx 缺签名头放行', () => {
    const verifier = new ResponseVerifier({ platformAuthPublicKey: platformAuth.publicKeyPem });
    for (const status of [200, 201, 204]) {
      expect(() =>
        verifier.verify({
          requestCanonicalSha256: vector.requestCanonicalSha256,
          apiVersion: '1',
          externalPath: '/x',
          status,
          headers: {},
          body: null,
        }),
      ).toThrowError(PlutusSignatureError);
    }

    for (const status of [401, 403, 404, 429, 500]) {
      const passed = verifier.verify({
        requestCanonicalSha256: vector.requestCanonicalSha256,
        apiVersion: '1',
        externalPath: '/x',
        status,
        headers: {},
        body: null,
      });
      expect(passed.verified).toBe(false);
      expect(passed.canonicalString).toBeNull();
    }
  });

  it('requireSignatureOnErrorResponses: true 时非 2xx 缺签名头也抛错', () => {
    const verifier = new ResponseVerifier({
      platformAuthPublicKey: platformAuth.publicKeyPem,
      requireSignatureOnErrorResponses: true,
    });
    for (const status of [401, 500]) {
      expect(() =>
        verifier.verify({
          requestCanonicalSha256: vector.requestCanonicalSha256,
          apiVersion: '1',
          externalPath: '/x',
          status,
          headers: {},
          body: null,
        }),
      ).toThrowError(PlutusSignatureError);
    }
  });
});

describe('密钥指纹', () => {
  it('四把密钥的指纹与向量一致,私钥推导结果相同', async () => {
    const { keyFingerprint, loadPrivateKey, loadPublicKey } = await import('../src/index.js');
    for (const name of ['merchant_auth', 'platform_auth', 'merchant_enc', 'platform_enc'] as const) {
      const pair = key(name);
      expect(keyFingerprint(loadPublicKey(pair.publicKeyPem))).toBe(pair.fingerprint);
      expect(keyFingerprint(loadPrivateKey(pair.privateKeyPem))).toBe(pair.fingerprint);
      expect(pair.publicExponent).toBe(65537);
      expect(pair.modulusBits).toBe(2048);
    }
  });

  it('指纹是 SHA256: 前缀加小写 hex', () => {
    const der = Buffer.from(
      key('platform_auth').publicKeyPem
        .split('\n')
        .filter((line) => !line.startsWith('-----'))
        .join(''),
      'base64',
    );
    expect(key('platform_auth').fingerprint).toBe(`SHA256:${createHash('sha256').update(der).digest('hex')}`);
  });
});
